File: src/SkyBlock/tiles/AutoMinerTile.php
<?php


namespace SkyBlock\tiles;


use pocketmine\block\Block;
use pocketmine\block\BlockTypeIds;
use pocketmine\block\tile\Chest;
use pocketmine\block\tile\Nameable;
use pocketmine\block\tile\NameableTrait;
use pocketmine\block\tile\Spawnable;
use pocketmine\block\VanillaBlocks;
use pocketmine\data\bedrock\EnchantmentIdMap;
use pocketmine\data\bedrock\EnchantmentIds;
use pocketmine\item\enchantment\EnchantmentInstance;
use pocketmine\item\Item;
use pocketmine\item\VanillaItems;
use pocketmine\math\Facing;
use pocketmine\nbt\tag\CompoundTag;
use pocketmine\player\Player;
use pocketmine\Server;
use SkyBlock\block\AutoMiner;
use SkyBlock\command\Functions;
use SkyBlock\Main;
use function array_key_exists;

class AutoMinerTile extends Spawnable implements Nameable {
    use NameableTrait;

    public const TAG_LEVEL = "Level";
    public const TAG_FORTUNE = "Fortune";
    public const TAG_FORTUNE_LEVEL = "Fortune_Level";
    public const MAX_LEVEL = 4;
    public const MAX_FORTUNE_LEVEL = 15;

    /** @var int */
    public int $delay = 200, $level1 = 0, $fortune = 0, $flevel = 1;
    public array $blocks
        = [
            BlockTypeIds::DIAMOND_ORE                => true,
            BlockTypeIds::GOLD_ORE                   => true,
            BlockTypeIds::EMERALD_ORE                => true,
            BlockTypeIds::NETHERRACK                 => true,
            BlockTypeIds::LAPIS_LAZULI_ORE           => true,
            BlockTypeIds::IRON_ORE                   => true,
            BlockTypeIds::COPPER_ORE                 => true,
            BlockTypeIds::COAL_ORE                   => true,
            BlockTypeIds::COBBLESTONE                => true,
            BlockTypeIds::NETHER_QUARTZ_ORE          => true,
            BlockTypeIds::QUARTZ                     => true,
            BlockTypeIds::DEEPSLATE_COAL_ORE         => true,
            BlockTypeIds::DEEPSLATE_COPPER_ORE       => true,
            BlockTypeIds::DEEPSLATE_IRON_ORE         => true,
            BlockTypeIds::DEEPSLATE_GOLD_ORE         => true,
            BlockTypeIds::DEEPSLATE_LAPIS_LAZULI_ORE => true,
            BlockTypeIds::DEEPSLATE_DIAMOND_ORE      => true,
            BlockTypeIds::DEEPSLATE_EMERALD_ORE      => true,
            BlockTypeIds::ANCIENT_DEBRIS             => true,
            BlockTypeIds::NETHERITE                  => true,
        ];

    public function onUpdate() : bool {
        if ($this->level1 === 0) $this->close();
        if ($this->closed) return false;
        $block = $this->getBlock();
        if (!$block instanceof AutoMiner) return false;
        $inst = Main::getInstance();
        if (!isset($inst->autominer[$this->getPosition()->getWorld()->getDisplayName()])) {
            if ($this->delay <= 0) {
                $this->delay = $this->getDelayByLevel($this->level1);
                $down = $this->getPosition()->getWorld()->getBlock($this->getBlock()->getSide(Facing::DOWN)->getPosition()->add(0, -1, 0));
                $up = $this->getPosition()->getWorld()->getTile($this->getBlock()->getSide(Facing::UP)->getPosition());
                if (array_key_exists($down->getTypeId(), $this->blocks) and $up instanceof Chest) {
                    $item = VanillaItems::DIAMOND_PICKAXE();
                    if ($this->fortune == 1) {
                        if ($this->flevel > self::MAX_FORTUNE_LEVEL) $this->flevel = self::MAX_FORTUNE_LEVEL;
                        if ($this->flevel < 1) $this->flevel = 1;
                        $item->addEnchantment(new EnchantmentInstance(EnchantmentIdMap::getInstance()->fromId(EnchantmentIds::FORTUNE), $this->flevel));
                    }
                    $affectedBlocks = $down->getAffectedBlocks();
                    $drops = array_merge(...array_map(function(Block $down) use ($item) : array {
                        return $down->getDrops($item);
                    }, $affectedBlocks
                                            )
                    );
                    $inv = $up->getInventory();
                    $smelt = [
                        BlockTypeIds::GOLD_ORE             => VanillaItems::GOLD_INGOT(),
                        BlockTypeIds::IRON_ORE             => VanillaItems::IRON_INGOT(),
                        BlockTypeIds::COPPER_ORE           => VanillaItems::COPPER_INGOT(),
                        BlockTypeIds::DEEPSLATE_GOLD_ORE   => VanillaItems::GOLD_INGOT(),
                        BlockTypeIds::DEEPSLATE_IRON_ORE   => VanillaItems::IRON_INGOT(),
                        BlockTypeIds::DEEPSLATE_COPPER_ORE => VanillaItems::COPPER_INGOT()
                    ];
                    foreach ($drops as $drop) {
                        /** @var Item $drop */
                        if ($this->fortune > 1 and $drop->getTypeId() != VanillaBlocks::COBBLESTONE()->asItem()->getTypeId() and $drop->getTypeId() != VanillaBlocks::ANCIENT_DEBRIS()->asItem()->getTypeId() and $drop->getTypeId() != VanillaBlocks::NETHERITE()->asItem()->getTypeId()) {
                            if (!str_contains(strtolower($block->getName()), "deepslate")) {
                                $count = mt_rand(1, $this->fortune);
                                $drop->setCount($count);
                            } else {
                                /** fortune item drops for normal ores and deepslate ores done here */
                                $count = mt_rand(1, $this->fortune);
                                if ($drop->getTypeId() === BlockTypeIds::QUARTZ) {
                                    $count = $count * 4;
                                    $drop = VanillaItems::NETHER_QUARTZ();
                                }
                                $drop->setCount((int) ceil($count * 1.8));
                            }
                        }
                        if (isset($smelt[$drop->getTypeId()])) $drop = $smelt[$drop->getTypeId()]->setCount($drop->getCount());
                        if ($inv->canAddItem($drop)) {
                            $inv->addItem($drop);
                        } else {
                            return true;
                        }
                    }
                    $this->getPosition()->getWorld()->setBlock($down->getPosition()->asVector3(), VanillaBlocks::AIR());
                    $blocks = [
                        BlockTypeIds::COBBLESTONE       => 0,
                        BlockTypeIds::COAL_ORE          => 1,
                        BlockTypeIds::COPPER_ORE        => 2,
                        BlockTypeIds::IRON_ORE          => 3,
                        BlockTypeIds::LAPIS_LAZULI_ORE  => 4,
                        BlockTypeIds::GOLD_ORE          => 5,
                        BlockTypeIds::DIAMOND_ORE       => 6,
                        BlockTypeIds::EMERALD_ORE       => 7,
                        BlockTypeIds::NETHER_QUARTZ_ORE => 8,
                        BlockTypeIds::ANCIENT_DEBRIS    => 9,

                        BlockTypeIds::DEEPSLATE_COAL_ORE         => 10,
                        BlockTypeIds::DEEPSLATE_COPPER_ORE       => 11,
                        BlockTypeIds::DEEPSLATE_IRON_ORE         => 12,
                        BlockTypeIds::DEEPSLATE_LAPIS_LAZULI_ORE => 13,
                        BlockTypeIds::DEEPSLATE_GOLD_ORE         => 14,
                        BlockTypeIds::DEEPSLATE_DIAMOND_ORE      => 15,
                        BlockTypeIds::DEEPSLATE_EMERALD_ORE      => 16,
                        BlockTypeIds::QUARTZ                     => 17,
                        BlockTypeIds::NETHERITE                  => 18,
                    ];
                    if (isset($blocks[$down->getTypeId()])) {
                        if (($island = $inst->getIslandManager()->getOnlineIslandByWorld($this->getPosition()->getWorld()->getDisplayName())) !== null) {
                            $receiver = $island->getReceiver();
                            $player = Server::getInstance()->getPlayerExact($receiver);
                            if (!$player instanceof Player) {
                                if (($player = $island->getRandomOnlineCoOwner()) === null) return true;
                                $receiver = $player->getName();
                            }
                            if (($user = $inst->getUserManager()->getOnlineUser($receiver)) !== null && $island->isAnOwner($receiver)) {
                                $user->addMana($blocks[$down->getTypeId()]);
                                $island->setPoints($blocks[$down->getTypeId()]);
                                Functions::safeXPAdd($user, $down->getXpDropForTool($item));
                            }
                        }
                    }
                }

            } else {
                --$this->delay;
            }
        }
        // main code goes here
        return true;
    }

    public function getName() : string {
        return "AutoMiner";
    }

    public function getDefaultName() : string {
        return "AutoMiner";
    }

    public function readSaveData(CompoundTag $nbt) : void {
        if ($nbt->getInt(self::TAG_LEVEL, null) === null) $this->close();

        $this->level1 = $nbt->getInt(self::TAG_LEVEL, $this->level1);
        $this->fortune = $nbt->getInt(self::TAG_FORTUNE, $this->fortune);
        $this->flevel = $nbt->getInt(self::TAG_FORTUNE_LEVEL, $this->flevel);

        $this->delay = $this->getDelayByLevel($this->level1);
        //$this->scheduleUpdate(); // schedules onUpdate
    }

    public function getDelayInSeconds() : int {
        return (int) ($this->getDelayByLevel($this->level1) / 20);
    }

    public function getDelayByLevel(int $level) : int {
        if ($level == 1) return 15 * 20; // 15 secs
        elseif ($level == 2) return 10 * 20;
        elseif ($level == 3) return 8 * 20;
        elseif ($level == 4) return 4 * 20;
        else return 200;
    }

    public function setDelay(int $delay) : void {
        $this->delay = $delay;
    }

    protected function writeSaveData(CompoundTag $nbt) : void {
        $this->applyBaseNBT($nbt);
    }

    private function applyBaseNBT(CompoundTag $nbt) : void {
        $nbt->setInt(self::TAG_LEVEL, $this->level1);
        $nbt->setInt(self::TAG_FORTUNE, $this->fortune);
        $nbt->setInt(self::TAG_FORTUNE_LEVEL, $this->flevel);
    }

    protected function addAdditionalSpawnData(CompoundTag $nbt) : void {
        $this->applyBaseNBT($nbt);
    }

}

File: SkyblockCore/src/main/java/dev/skyblock/tiles/AutoMinerTile.java
package dev.skyblock.tiles;

import dev.skyblock.SkyblockCore;
import dev.skyblock.island.Island;
import dev.skyblock.user.User;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.Chest;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class AutoMinerTile extends BaseTile {
    public static final int MAX_LEVEL = 4;
    public static final int MAX_FORTUNE_LEVEL = 15;

    private int tickCounter = 0;
    private final TileData data;

    // Block material → mana value (matches PHP source exactly)
    private static final Map<Material, Integer> MANA_VALUES = new HashMap<>();
    static {
        MANA_VALUES.put(Material.COBBLESTONE, 0);
        MANA_VALUES.put(Material.COAL_ORE, 1);
        MANA_VALUES.put(Material.COPPER_ORE, 2);
        MANA_VALUES.put(Material.IRON_ORE, 3);
        MANA_VALUES.put(Material.LAPIS_ORE, 4);
        MANA_VALUES.put(Material.GOLD_ORE, 5);
        MANA_VALUES.put(Material.DIAMOND_ORE, 6);
        MANA_VALUES.put(Material.EMERALD_ORE, 7);
        MANA_VALUES.put(Material.NETHER_QUARTZ_ORE, 8);
        MANA_VALUES.put(Material.ANCIENT_DEBRIS, 9);
        MANA_VALUES.put(Material.DEEPSLATE_COAL_ORE, 10);
        MANA_VALUES.put(Material.DEEPSLATE_COPPER_ORE, 11);
        MANA_VALUES.put(Material.DEEPSLATE_IRON_ORE, 12);
        MANA_VALUES.put(Material.DEEPSLATE_LAPIS_ORE, 13);
        MANA_VALUES.put(Material.DEEPSLATE_GOLD_ORE, 14);
        MANA_VALUES.put(Material.DEEPSLATE_DIAMOND_ORE, 15);
        MANA_VALUES.put(Material.DEEPSLATE_EMERALD_ORE, 16);
        MANA_VALUES.put(Material.QUARTZ_BLOCK, 17);
        MANA_VALUES.put(Material.NETHERITE_BLOCK, 18);
    }

    public AutoMinerTile(Location location, TileData data) {
        super(location);
        this.data = data;
        this.tickCounter = getDelayTicks(data.getLevel());
    }

    @Override
    public void onUpdate() {
        if (tickCounter > 0) {
            tickCounter--;
            return;
        }
        tickCounter = getDelayTicks(data.getLevel());

        SkyblockCore plugin = SkyblockCore.getInstance();
        String worldName = location.getWorld().getName();

        Island island = plugin.getIslandManager().getOnlineIslandByWorld(worldName).orElse(null);
        if (island == null) return;

        // Block to mine = 2 below the AutoMiner
        Block target = location.getBlock().getRelative(0, -2, 0);
        // Chest = 1 above the AutoMiner
        Block aboveBlock = location.getBlock().getRelative(0, 1, 0);

        if (!MANA_VALUES.containsKey(target.getType())) return;
        if (!(aboveBlock.getState() instanceof Chest chestState)) return;

        Inventory inv = chestState.getInventory();
        if (inv.firstEmpty() == -1) return; // chest full, skip

        // Build a pickaxe with fortune if enabled
        ItemStack pickaxe = new ItemStack(Material.DIAMOND_PICKAXE);
        if (data.getFortuneEnabled() == 1) {
            int fl = Math.min(Math.max(data.getFortuneLevel(), 1), MAX_FORTUNE_LEVEL);
            pickaxe.addUnsafeEnchantment(Enchantment.FORTUNE, fl);
        }

        // Get drops and add to chest
        Collection<ItemStack> drops = target.getDrops(pickaxe);
        for (ItemStack drop : drops) {
            inv.addItem(drop);
        }

        // Award mana and island points to receiver
        String receiverName = island.getReceiver();
        Player receiver = plugin.getServer().getPlayerExact(receiverName);
        if (receiver == null) receiver = island.getRandomOnlineCoOwner();
        if (receiver != null) {
            User user = plugin.getUserManager().getOnlineUser(receiver.getUniqueId());
            if (user != null && island.isAnOwner(receiver.getName())) {
                int manaGain = MANA_VALUES.getOrDefault(target.getType(), 0);
                user.addMana(manaGain);
                island.setPoints(island.getPoints() + manaGain);
            }
        }

        // Remove the mined block
        target.setType(Material.AIR);
    }

    public static int getDelayTicks(int level) {
        return switch (level) {
            case 1 -> 15 * 20;
            case 2 -> 10 * 20;
            case 3 -> 8 * 20;
            case 4 -> 4 * 20;
            default -> 200;
        };
    }

    public TileData getData() { return data; }
}
